File: src/main/java/org/embulk/test/Recorder.java
package org.embulk.test;

import com.google.common.collect.ImmutableList;
import org.embulk.spi.Column;
import org.embulk.spi.PageReader;
import org.embulk.spi.Schema;
import org.embulk.spi.util.Pages;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertThat;

/**
 * Repository of {@link Record}
 */
class Recorder {
    private List<Record> records = new ArrayList<>();
    private Schema schema;

    Recorder() {
    }

    private List<Record> getRecords() {
        return new ArrayList<>(this.records);
    }

    synchronized void addRecord(PageReader reader) {
        final List<Object> values = new ArrayList<>();
        reader.getSchema().visitColumns(new Pages.ObjectColumnVisitor(reader) {
            @Override
            public void visit(Column column, Object value) {
                values.add(value);
            }
        });
        this.records.add(new Record(values));
    }

    synchronized void setSchema(Schema schema) {
        this.schema = schema;
    }

    void clear() {
        this.records = new ArrayList<>();
        this.schema = null;
    }

    void assertRecords(Record... records) {
        Set<Record> actual = new HashSet<>();
        for (Record record : getRecords()) {
            actual.add(record);
        }

        Set<Record> expected = new HashSet<>();
        Collections.addAll(expected, records);

        assertThat(actual, is(expected));
    }

    void assertSchema(Column... columns) {
        Schema.Builder builder = Schema.builder();
        for (Column column : columns) {
            builder.add(column.getName(), column.getType());
        }
        assertThat(builder.build(), is(this.schema));
    }
}

File: src/main/kotlin/org/embulk/test/Recorder.kt
package org.embulk.test

import org.embulk.spi.Column
import org.embulk.spi.Schema
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.`is`

/**
 * Repository of [Record]
 */
internal class Recorder {
    private val records = mutableListOf<Record>()
    private var schema: Schema? = null

    @Synchronized fun addRecord(record: Record) {
        this.records.add(record)
    }

    @Synchronized fun setSchema(schema: Schema) {
        this.schema = schema
    }

    @Synchronized fun clear() {
        this.records.clear()
        this.schema = null
    }

    fun assertRecords(vararg records: Record) {
        assertThat(this.records.toHashSet(), `is`(records.toHashSet()))
    }

    fun assertSchema(vararg columns: Column) {
        val builder = columns.fold(Schema.builder()) { builder, column -> builder.add(column.name, column.type) }
        assertThat(builder.build(), `is`(this.schema))
    }
}
